lib/coroutine.php

```diff
@@ -88,17 +85,31 @@ function(Throwable $reason) use ($generator, $advanceGenerator) {
                 break;
             } else {
                 // If the value was not a promise, we'll just let it pass through.
-                $lastYieldResult = $yieldedValue;
                 $generator->send($yieldedValue);
             }
 
         }
 
         // If the generator is at the end, and we didn't run into an exception,
-        // we can fullfill the promise with the last thing that was yielded to
-        // us.
-        if (!$generator->valid() && $promise->state === Promise::PENDING) {
-            $promise->fulfill($lastYieldResult);
+        // We're grabbing the "return" value and fulfilling our top-level
+        // promise with its value.
+         if (!$generator->valid() && $promise->state === Promise::PENDING) {
+             $returnValue = $generator->getReturn();
```

Comment:
That comment is incorrect. `getReturn()` will work as long as the generator has _returned_. This includes having reached the end of the function.